### PR TITLE
Offers in autocomplete

### PR DESCRIPTION
**Pull request by martaswiatkowska**
Added offers to autocomplete results.
After click on displayed offers users are redirected to offer on service page by anchor.
Only published offers are indexed and if the service has more than two offers.

### CODE REVIEW

File: app/controllers/concerns/service/autocomplete.rb [hunk 1]
```diff
@@ -21,12 +23,33 @@ def autocomplete
 
   private
     def generate_html(query)
-      html_results = query.with_highlights.map { |s, h| "<li class=\"dropdown-item\" role=\"option\" data-autocomplete-value=\"#{s[:id]}\">#{h[:title]}</li>" }.join
+      service_results = "<p class=\"list-header\"> Services </p>"
```
Comment (mkasztelnik) [inline app/controllers/concerns/service/autocomplete.rb]: This generation starts to be complex. Maybe a good solution would be to move HTML generation to `haml`?

File: app/models/offer.rb [hunk 2]
```diff
@@ -3,11 +3,27 @@
...
     published: "published",
     draft: "draft"
   }
 
+  def search_data
+    {
+      offer_id: id,
+      name: name,
+      description: description,
+      service_id: service_id
```
Comment (mkasztelnik) [inline app/models/offer.rb]: Can we add a published flag to elasticsearch and use it to limit offers found in MP (autocomplete in backoffice should search in all offers).

File: app/helpers/service_helper.rb [hunk 3]
```diff
@@ -59,4 +59,8 @@ def service_logo(service)
       image_tag "eosc-img.png"
     end
   end
+
+  def get_service_title(id)
+    Service.find(id).title
```
Comment (mkasztelnik) [inline app/helpers/service_helper.rb]: This is quite a danger and can lead to `n+1` queries problem. Can we use a mechanism similar to service search `scope_results` to have offers already joined with services?

File: app/views/services/autocomplete/_offer_item.html.haml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+%li.dropdown-item{ "role": "option",
+  "data-autocomplete-value": "#{offer.service_id}",
+  "data-autocomplete-anchor": "#{offer.id}" }
+  %span= get_service_title(offer.service_id)
```
Comment (mkasztelnik) [inline app/views/services/autocomplete/_offer_item.html.haml]: `span` should be used only for the service title of for the whole line?

Comment (mkasztelnik) [inline app/views/services/autocomplete/_offer_item.html.haml]: Can we use joins in the offers result and `offer.service.title`?

File: app/models/offer.rb (hunk 2 above)
Comment (martaswiatkowska) [reply app/models/offer.rb]: Function should_index? are filtering published offers https://github.com/cyfronet-fid/marketplace/pull/1420/files/e5d87a8fb986d2c019adde45f32c1be5a3fb4c44#diff-a1928e578416256fe9dc55132da41a2bR23

Now I think we don't need offer_id and filtering by offer_id in request to elasticsearch.

File: app/views/services/autocomplete/_offer_item.html.haml (hunk 4 above)
Comment (martaswiatkowska) [reply app/views/services/autocomplete/_offer_item.html.haml]: Only for the service title because its element with different collor

**Comment by martaswiatkowska** [comment]
@wziajka doesn't have to do sth additional because already all models with searchkick are indexed by ```rake searchkick:reindex:all```

File: app/models/offer.rb (hunk 2 above)
Comment (mkasztelnik) [reply app/models/offer.rb]: Ok, it is good for now. When the offer search will be added to the backoffice we will change this mechanism and only then add this published flag.

**Pull request merged**